### PR TITLE
SCP-4277 Contract ID management

### PR DESCRIPTION
**Pull request by jhbertra**
Changes included:

- Adds a new `marlowe-history` server process which manages multiple follower threads and exposes a query API and a command API (as well as a temporary repl)
- Adds a new `marlowe-history-cli` which can be used to communicate with a `marlowe-history` server. supported commands:
  - `marlowe-history-cli add CONTRACT_ID` starts following a contract by its ID
  - `marlowe-history-cli rm CONTRACT_ID` stops following a contract by its ID
  - `marlowe-history-cli ls -[sf]` lists all followed contracts. `-s` shows the status of the follower and `-f` also shows failed contracts (e.g. if you use `add` and it fails to follow the contract, the failure message will be shown next to the contract ID with `-sf`)

### CODE REVIEW

File: marlowe-runtime/marlowe-history-cli/Main.hs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    printContractId = T.putStrLn . renderContractId
+    printStatus contractId status = do
+      T.putStr $ renderContractId contractId
+      putStr " Status: "
+      print status
+
+failWith :: String -> IO a
+failWith msg = do
+  hPutStrLn stderr msg
+  exitFailure
```
Comment (bwbush) [inline marlowe-runtime/marlowe-history-cli/Main.hs:93]: This function is identical to `System.Exit.die` in `base`.

File: marlowe-runtime/marlowe-history-cli/Main.hs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+getOptions :: IO Options
+getOptions = O.execParser $ O.info parser infoMod
+  where
+    parser = O.helper <*> (Options <$> commandPortParser <*> queryPortParser <*> hostParser <*> commandParser)
+    commandPortParser = O.option O.auto $ mconcat
+      [ O.long "command-port"
+      , O.short 'p'
+      , O.value 3717
+      , O.metavar "PORT_NUMBER"
+      , O.help "The port number for issuing commands to the history server"
```
Comment (bwbush) [inline marlowe-runtime/marlowe-history-cli/Main.hs:129]: Should we list the default port numbers in the help strings?

**Review by bwbush — CHANGES_REQUESTED** [review]
Looks good, and `chainseekd`, `marlowe-history`, `marlowe-history-cli`, and `marlowe-follower` all seem to be working in concert for me on the `preview` network, where I tested them with contract `75601b84f1581da4d4c3f1ac3818cee89ce674f288ee152b24938555e982e0fe#1`.

One of the commands is missing help
```
cabal run exe:marlowe-history -- --help                                                                                                                                                                                     
Up to date
Invalid option `--help'

Usage: marlowe-history [--chain-seek-port-number PORT_NUMBER] 
                       [--chain-seek-query-port-number PORT_NUMBER] 
                       [-c|--command-port PORT_NUMBER] 
                       [-c|--query-port PORT_NUMBER] 
                       [--chain-seek-host HOST_NAME] [-h|--host HOST_NAME]

  Contract follower for Marlowe Runtime
```

I also noticed two `-p` options in `chainseekd`:
```
Usage: chainseekd [--version] (-s|--socket-path SOCKET_FILE) 
                  [-m|--testnet-magic INTEGER] [-d|--database-uri DATABASE_URI]
                  --genesis-config-file-hash CONFIG_HASH
                  --genesis-config-file CONFIG_FILE [-h|--host HOST_NAME] 
                  [-p|--port-number PORT_NUMBER] 
                  [-p|--query-port-number PORT_NUMBER]
```
The same problem happens for `-p` and -`c` in `marlowe-history` and `marlowe-history-cli` and `marlowe-follower`. We should probably review all of the CLIs and guard against the root cause.

**Comment by jhbertra** [comment]
> We should probably review all of the CLIs and guard against the root cause.

The root cause being copy-pasting code? Not sure how we're going to guard against that in practice :stuck_out_tongue: I'm reviewing all the CLI options now

**Review by bwbush — APPROVED** [review]

**Pull request merged**